File: greedy/134.gas-station.py
#
# @lc app=leetcode id=134 lang=python3
#
# [134] Gas Station
#

# @lc code=start
class Solution:
    def canCompleteCircuit(self, gas: List[int], cost: List[int]) -> int:
        start = 0
        totalSum = 0
        curSum = 0

        for i in range(len(gas)):
            rest = gas[i] - cost[i]
            totalSum += rest
            curSum += rest
            if curSum < 0:
                start = i + 1
                curSum = 0

        if totalSum < 0:
            return -1
        return start
# @lc code=end

File: Greedy/134.gas-station.java
/*
 * @lc app=leetcode id=134 lang=java
 *
 * [134] Gas Station
 */

// @lc code=start
class Solution {
    public int canCompleteCircuit(int[] gas, int[] cost) {
        int n = gas.length;
        int totalDiff = 0;
        int curSum = 0;
        int res = 0;

        for (int i = 0; i < n; i++) {
            int remainGas = gas[i] - cost[i];
            totalDiff += remainGas;
            curSum += remainGas;
            if (curSum < 0) {
                res = i + 1;
                curSum = 0;
            }
        }

        if (totalDiff < 0) {
            return -1;
        }
        return res;
    }
}
// @lc code=end
